### PR TITLE
Forms

### PR DESCRIPTION
**Pull request by Licini**
1. A simple universal Add form for creating primitives/shapes in viewer. A corresponding button for added for each `Object` that has a `create_default` function
image

2. Improved Edit form layout using expandable boxes, and will appear as integrated dock widget (which can also be detached from main window)
image

3. Nice COMPAS application logo ; )
image

### CODE REVIEW

**Comment by brgcode** [comment]
i prefer option 1

**Review by brgcode — APPROVED** [review]
a few small things here and there as marked in the comments...

also, are the schemas used anywhere for validation?

perhaps we should refine `create_default` for some of the objects. for example, the current version does not make much sense for a line i find. we should also consider to create a frame and not a plane for most of the primitives and shapes, but i guess that is not the concern of this PR...

would also make sense to access this functionality from a toolbar with icons (to be discussed with juney)...

File: src/compas_view2/app/config.json [hunk 1]
```diff
@@ -68,26 +68,7 @@
...
-                    "type": "menu",
-                    "text": "Shapes",
-                    "items":
-                    [
-                        {"type": "action", "text": "Add Box", "action": "add_box"},
-                        {"type": "action", "text": "Add Sphere", "action": "add_sphere"},
-                        {"type": "action", "text": "Add Torus", "action": "add_torus"}
-                    ]
-                },
+                {"type": "action", "text": "Add Object", "action": "add_object"},
```
[the remark this answers is not in the document]
Comment (Licini) [reply src/compas_view2/app/config.json:71]: sure, will make the change, how about the command text? Should it also be ‘Add Geometry Object’ ?

File: src/compas_view2/app/controller.py [hunk 2]
```diff
@@ -170,6 +174,13 @@ def history(self):
     # Primitive actions
     # ==============================================================================
 
+    def add_object(self):
```
[the remark this answers is not in the document]
Comment (Licini) [reply src/compas_view2/app/controller.py:177]: Ah yes forgot about those, will clean them up

File: src/compas_view2/forms/addform.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+from PySide2 import QtWidgets
+
+from .form import Form
+from ..objects.object import DATA_VIEW
```
[the remark this answers is not in the document]
Comment (Licini) [reply src/compas_view2/forms/addform.py]: yes renamed now

File: src/compas_view2/forms/collapsiblebox.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        del lay
+        self.content_area.setLayout(layout)
+        self._collapsed_height = self.sizeHint().height()
+
+    def expand(self):
+        self.toggle_button.setArrowType(QtCore.Qt.DownArrow)
+        self.setFixedHeight(self._collapsed_height + self.get_content_height())
+        self.content_area.setMaximumHeight(self.get_content_height())
+
+    def contract(self):
```
[the remark this answers is not in the document]
Comment (Licini) [reply src/compas_view2/forms/collapsiblebox.py]: done

File: src/compas_view2/forms/collapsiblebox.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        self.toggle_button.setArrowType(QtCore.Qt.RightArrow)
+        self.setFixedHeight(self._collapsed_height)
+        self.content_area.setMaximumHeight(0)
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (Licini) [reply src/compas_view2/forms/collapsiblebox.py:68]: yep but actually not needed anymore. I just removed it

File: src/compas_view2/forms/collapsiblebox.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            self.collapse()
+        self.update_parent()
+
+    def get_content_height(self):
+        return self._layout.sizeHint().height()
+
+    def setContentLayout(self, layout):
+        self._layout = layout
+        lay = self.content_area.layout()
+        del lay
```
Comment (brgcode) [inline src/compas_view2/forms/collapsiblebox.py:53]: ?

File: src/compas_view2/objects/object.py [hunk 7]
```diff
@@ -37,13 +36,13 @@ class Object(ABC):
     @staticmethod
     def register(dtype, vtype):
```
Comment (brgcode) [inline src/compas_view2/objects/object.py:37]: `vtype` => `otype`

**Review by brgcode — APPROVED** [review]

File: src/compas_view2/forms/collapsiblebox.py (hunk 6 above)
Comment (Licini) [reply src/compas_view2/forms/collapsiblebox.py]: was left over from example I looked up for, seems no use, now deleted

File: src/compas_view2/objects/object.py (hunk 7 above)
Comment (Licini) [reply src/compas_view2/objects/object.py]: done, gonna merge now

**Pull request merged**